hwy/base.h

```diff
@@ -980,6 +636,613 @@ using RemoveRef = typename RemoveRefT<T>::type;
+  };
+};
+
+#if HWY_COMPILER_MSVC
+HWY_DIAGNOSTICS(pop)
+#endif
+
+template <class From, class To>
+HWY_API constexpr bool IsConvertible() {
+  return IsConvertibleT<From, To>::value;
+}
+
+template <class From, class To>
+class IsStaticCastableT {
+ private:
+  template <class T, class U, class = decltype(static_cast<U>(DeclVal<T>()))>
+  static hwy::SizeTag<1> TryStaticCastTest(int);
+
+  template <class T, class U, class Arg>
+  static hwy::SizeTag<0> TryStaticCastTest(Arg);
+
+ public:
+  enum {
+    value = IsSame<decltype(TryStaticCastTest<From, To>(0)), hwy::SizeTag<1>>()
+  };
+};
+
+template <class From, class To>
+static constexpr bool IsStaticCastable() {
+  return IsStaticCastableT<From, To>::value;
+}
+
+template <class T, class From>
+class IsAssignableT {
+ private:
+  template <class T1, class T2, class = decltype(DeclVal<T1>() = DeclVal<T2>())>
+  static hwy::SizeTag<1> TryAssignTest(int);
+
+  template <class T1, class T2, class Arg>
+  static hwy::SizeTag<0> TryAssignTest(Arg);
+
+ public:
+  enum {
+    value = IsSame<decltype(TryAssignTest<T, From>(0)), hwy::SizeTag<1>>()
+  };
+};
+
+template <class T, class From>
+static constexpr bool IsAssignable() {
+  return IsAssignableT<T, From>::value;
+}
+
+//------------------------------------------------------------------------------
+// F16/BF16 lane types
+
+#pragma pack(push, 1)
+
+// float16_t load/store/conversion intrinsics are always supported on Armv8 and
+// VFPv4 (except with MSVC). On Armv7 Clang requires __ARM_FP & 2; GCC requires
+// -mfp16-format=ieee.
+#if (HWY_ARCH_ARM_A64 && !HWY_COMPILER_MSVC) ||                    \
+    (HWY_COMPILER_CLANG && defined(__ARM_FP) && (__ARM_FP & 2)) || \
+    (HWY_COMPILER_GCC_ACTUAL && defined(__ARM_FP16_FORMAT_IEEE))
+#define HWY_NEON_HAVE_FLOAT16C 1
+#else
+#define HWY_NEON_HAVE_FLOAT16C 0
+#endif
+
+// C11 extension ISO/IEC TS 18661-3:2015 but not supported on all targets.
+// Required if HWY_HAVE_FLOAT16, i.e. RVV with zvfh or AVX3_SPR (with
+// sufficiently new compiler supporting avx512fp16). Do not use on clang-cl,
+// which is missing __extendhfsf2.
+#if ((HWY_ARCH_RVV && defined(__riscv_zvfh) && HWY_COMPILER_CLANG) || \
+     (HWY_ARCH_X86 && defined(__SSE2__) &&                            \
+      ((HWY_COMPILER_CLANG >= 1600 && !HWY_COMPILER_CLANGCL) ||       \
+       HWY_COMPILER_GCC_ACTUAL >= 1200)))
+#define HWY_HAVE_C11_FLOAT16 1
+#else
+#define HWY_HAVE_C11_FLOAT16 0
+#endif
+
+// If 1, both __bf16 and a limited set of *_bf16 SVE intrinsics are available:
+// create/get/set/dup, ld/st, sel, rev, trn, uzp, zip.
+#if HWY_ARCH_ARM_A64 && !(HWY_COMPILER_CLANG && HWY_COMPILER_CLANG < 1700) && \
+    (defined(__ARM_FEATURE_SVE_BF16) || HWY_COMPILER_GCC_ACTUAL >= 1100)
+#define HWY_ARM_HAVE_BFLOAT16 1
+#else
+#define HWY_ARM_HAVE_BFLOAT16 0
+#endif
+
+#if HWY_ARM_HAVE_BFLOAT16 && \
+    (defined(__ARM_FEATURE_SVE_BF16) || defined(__ARM_FEATURE_SVE))
+#define HWY_SVE_HAVE_BFLOAT16 1
```

Comment:
It is possible to make HWY_SVE_HAVE_BFLOAT16 be a per-target setting.

I have also updated hwy/ops/arm_sve-inl.h in this pull request to allow the SVE targets to use svbfloat16_t vectors if supported by the current C++ compiler, even when hwy::bfloat16_t is not a wrapper around the __bf16 type.

GCC 10 and later also supports the SVE svbfloat16_t type with the -march=armv8-a+sve option.

Reply:
Ah, so if I understand correctly, `HWY_SVE_HAVE_BFLOAT16_VEC` is the per-target setting that could also take effect without extra -march flags. It might help to add a comment at their definition explaining the difference between ARM_HAVE_BFLOAT16 and SVE_HAVE_BFLOAT16.

Reply:
> Ah, so if I understand correctly, `HWY_SVE_HAVE_BFLOAT16_VEC` is the per-target setting that could also take effect without extra -march flags. It might help to add a comment at their definition explaining the difference between ARM_HAVE_BFLOAT16 and SVE_HAVE_BFLOAT16.

The HWY_SVE_HAVE_BFLOAT16_VEC was added to allow the SVE svbfloat16_t type to be used if supported by the C++ compiler, even if hwy::bfloat16_t is not a wrapper around the __bf16 type.

It is also now possible for HWY_SVE_HAVE_BFLOAT16 to be 0 but for HWY_SVE_HAVE_BFLOAT16_VEC to be 1 if the SVE svbfloat16_t vector type is supported by the current C++ compiler. Allowing HWY_SVE_HAVE_BFLOAT16 to be 0 but for HWY_SVE_HAVE_BFLOAT16_VEC to be 1 works around compiler bugs that are there with the __bf16 type in GCC 10 and Clang 16 and earlier.

Reply:
Makes sense. These are great comments, and I'd love to see them in the source code because I will have forgotten it within 3 months. I can also add some comment after we merge if you prefer.